Adding LFP parsing by trials and lasers

bradlytstone opened the pull request:
Good luck

bradlytstone committed:
Adding LFP parsing by trials and lasers

narendramukherjee commented on LFP_analysis/LFP_Processing_Final.py:
```diff
@@ -0,0 +1,179 @@
+#Open the hdf5 file
+hf5 = tables.open_file(hdf5_name, 'r+')
+
+#Create vector of electodes
```
I would say: "Get the electrode numbers that have neurons on them from the unit_descriptor table"

narendramukherjee commented on LFP_analysis/LFP_Processing_Final.py:
```diff
@@ -0,0 +1,179 @@
+#Create vector of electodes
+electrodegroup = hf5.root.unit_descriptor[:]['electrode_number']
+
+#Remove duplicates within array
```
"Each electrode might have more than neuron, so will be repeated in the unit_descriptor table. So pick only unique electrode numbers"

narendramukherjee commented on LFP_analysis/LFP_Processing_Final.py:
```diff
@@ -0,0 +1,179 @@
+#Loop through each neuron-recording electrode, filter data, and create array in new LFP node
+for i in range(len(Raw_Electrodefiles)):
+
+    #Filter data
```
"Read data from .dat files as int16"

narendramukherjee commented on LFP_analysis/LFP_Processing_Final.py:
```diff
@@ -0,0 +1,179 @@
+
+    #Filter data
+    data = np.fromfile(Raw_Electrodefiles[i], dtype = np.dtype('int16'))
+    filt_el = get_filtered_electrode(data)
```
"Filter data"

narendramukherjee commented on LFP_analysis/LFP_Processing_Final.py:
```diff
@@ -0,0 +1,179 @@
+	#Delete data
+	hf5.remove_node('/raw_LFP', recursive = True)
+hf5.flush()
+
```
HDF5 files are weird in that just deleting a node often doesn't release actual space from the file. One needs to compress the file to actually release the space - maybe you can add a print statement here that tells the user to run blech_hdf5_repack.py (already in the blech_clust repo) that will compress their file and make everything nice :)

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,95 @@
+
+#Establish number of tastes and laser_combination variables
+taste_count = len(laser_nodes)
+laser_comb_count = len(laser_nodes)
```
len(laser_nodes) doesn't give you the count of laser combinations. Your code here will deduce that the number of laser combinations is the same as the number of tastes, which isn't the case generally. The laser combinations are contained in:
```
hf5.root.ancillary_analysis.laser_combination_d_l
```
Look that up!

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,95 @@
+#Establish number of tastes and laser_combination variables
+taste_count = len(laser_nodes)
+laser_comb_count = len(laser_nodes)
+trials = hf5.root.ancillary_analysis.trials[1,:]
```
What is this doing?

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,95 @@
+trials = hf5.root.ancillary_analysis.trials[1,:]
+trials_per_taste_las_cond = int(len(trials)/taste_count)
+
+def blockshaped(arr, nrows, ncols):
```
What's this function for?

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,95 @@
+    hf5.create_group('/LFP_Lasers', str.split(taste, '/')[-1])    
+    
+#Create loop based on number of laser conditions and create groupings of blank arrays to fill
+for z in range(taste_count):        #Where z indicates dig_in (laser condition)
```
I'm very confused about what's happening in the code from here on. I feel that this is far more complicated than it needs to be :'( I would basically suggest something along these lines:
```
# Get the laser conditions
laser_conditions = hf5.root.ancillary_analysis.laser_combination_d_l[:]
laser_trials = hf5.root.ancillary_analysis.trials[:]

# Get number of tastes
num_tastes = len(hf5.list_nodes('/spike_trains'))
# Get number of trials for each taste (assume equal number of trials for each)
num_trials = hf5.root.spike_trains.dig_in_0.spike_array[:].shape[....fill in the right numbers here]

# Run through the tastes and laser conditions, and fill up an array with the data
for i in num_tastes:
   for j in laser_conditions.shape[0]:
      # Pick the right trial numbers here, and fill them in your array
```
Basically, this part needs drastic modifications!

narendramukherjee reviewed (changes requested):
LFP_Processing_Final.py is phenomenally written, however the other piece of code needs a lot of modification. It is far too complicated than it needs to be. I have suggested a basic template/logic following which this code can be simply written, have a look at that, and let's discuss this again. Let's make modifications to the 2nd file before we merge.

Thanks for all your efforts, it looks amazing overall! :)

bradlytstone committed:
Updated "Laser_LFP_Parse_Final.py" with incorporated edits.
Revised "LFP_Processing_Final.py" to make it less complicated and run smoother.

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,69 @@
+for x in range(laser_conditions.shape[0]):
+    
+    #Create identifier based on laser combination information and create group within "LFP_Lasers" node
+    las_type = 'laser_combos_'+str(int(laser_conditions[x,0]))+'_'+str(int(laser_conditions[x,1]))         
```
Kinda forcing my terminology onto you at this point, but the reason I used d_l in the laser combo naming scheme I followed is so that people know that the first number is duration and the second number is lag. I think we should do that here too, so something like this will do:
```
las_type = 'laser_combos_d_l_'+str(int(laser_conditions[x,0]))+'_'+str(int(laser_conditions[x,1]))
```

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,69 @@
+    
+    #Create identifier based on laser combination information and create group within "LFP_Lasers" node
+    las_type = 'laser_combos_'+str(int(laser_conditions[x,0]))+'_'+str(int(laser_conditions[x,1]))         
+    hf5.create_group('/LFP_Lasers', str.split(las_type, '/')[-1])
```
Why are you doing:
```
str.split(las_type, '/')[-1]
```
Shouldn't this just be:
```
las_type
```
as is evidenced by your next piece of code a couple of lines later, which says:
```
hf5.create_array('/LFP_Lasers/%s' % las_type, '%s' %taste, np.asarray(final_array))
```

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,69 @@
+        taste = 'taste_' + str(y)
+        
+        #Target corresponding LFP array for data indexing
+        lfp_coll = np.mean(lfp_nodes[y][:],axis=(0))
```
Are we taking the mean across electrodes at this point? We should specify that in the comment then, so that the reader knows

narendramukherjee replied:
Also, if you are taking the mean only along one axis, you do not have to make the axis number a tuple (by putting it into brackets). This will do:
```
lfp_coll = np.mean(lfp_nodes[y][:],axis=0)
```

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,69 @@
+        start_trial = int(y*(num_trials/num_tastes))
+        trial_group = laser_trials[x,start_trial:int(start_trial+num_trials/num_tastes)]-int(y*num_trials) #Create last of trials to index based on the number of trials per taste (assumes equal trial numbers per tastant)
+        laser_trial_LFPs.append(lfp_coll[trial_group])
+        final_array = np.mean(laser_trial_LFPs,axis=(0))
```
Again, no need to put the axis parameter in as a tuple. Just an integer will do.

narendramukherjee replied:
What's this mean for? I thought we took a mean across electrodes already a few lines above.

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,69 @@
+        #Target corresponding LFP array for data indexing
+        lfp_coll = np.mean(lfp_nodes[y][:],axis=(0))
+        start_trial = int(y*(num_trials/num_tastes))
+        trial_group = laser_trials[x,start_trial:int(start_trial+num_trials/num_tastes)]-int(y*num_trials) #Create last of trials to index based on the number of trials per taste (assumes equal trial numbers per tastant)
```
I am not sure if these two lines actually pull out the trial numbers we want. Since this is critical, let's talk about this and you can explain what your logic was.

I would do something like:
```
# Pick the trials of taste y in laser condition x
trial_group = np.where((laser_trials[x, :] >= y*num_trials)*(laser_trials[x, :] < (y+1)*num_trials) == True)[0]

# The trials picked above are on the absolute scale of trial numbers (so taste 3, for instance, will have trials 90 to 119 if there were a total of 120 trials). We need to convert this into a 0-29 scale (for 30 trials of each taste) to be able to use these numbers on the LFP arrays
trial_group = laser_trials[x, trial_group] - int(y*num_trials)
```

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,69 @@
+        lfp_coll = np.mean(lfp_nodes[y][:],axis=(0))
+        start_trial = int(y*(num_trials/num_tastes))
+        trial_group = laser_trials[x,start_trial:int(start_trial+num_trials/num_tastes)]-int(y*num_trials) #Create last of trials to index based on the number of trials per taste (assumes equal trial numbers per tastant)
+        laser_trial_LFPs.append(lfp_coll[trial_group])
```
Why do we use append here? Why have laser_trial_LFPs be a list which you append to? Why not:
```
laser_trial_LFPs = lfp_coll[trial_group]
```

narendramukherjee commented on LFP_analysis/Laser_LFP_Parse_Final.py:
```diff
@@ -0,0 +1,69 @@
+        final_array = np.mean(laser_trial_LFPs,axis=(0))
+        
+        # Create arrays (based on tastants) to build under corresonding laser condition
+        hf5.create_array('/LFP_Lasers/%s' % las_type, '%s' %taste, np.asarray(final_array))
```
np.asarray will not be needed here if you didn't make the laser_trial_LFPs a list above.

narendramukherjee reviewed (changes requested):
Code 1 is perfect, and ready to be merged. Code 2 is so much improved :) However, there are some critical points in Code 2 that I am still confused about. Can you read through my comments today and see what you think (decide your responses etc)? Then you can give me those responses in person tomorrow - I might have misunderstood your logic at certain points, so you can explain that to me in response to the comments. Great work again :D

bradlytstone committed:
Addressed logic concerns and updated the code to be structurally more simplistic and added more comments for readability.

narendramukherjee reviewed (approved):
This looks great now, esp given you've checked that trials are pulled out correctly! Thanks for the contribution!
